pandas/core/indexes/datetimelike.py

```diff
@@ -1005,7 +1005,7 @@ def shift(self, n, freq=None):
             result = self + offset
 
             if hasattr(self, 'tz'):
```

Comment:
I generally [avoid `hasattr`](https://hynek.me/articles/hasattr/) in projects that support Python 2.

Reply:
yeah this impl is shared by DTI and TDI so this could be restructured a bit. Please file an issue .